docs/architecture/adr-016-validator-consensus-key-rotation.md

```diff
@@ -0,0 +1,109 @@
+# ADR 016: Validator Consensus Key Rotation
+
+## Changelog
+
+- 2019 Oct 23: Initial draft
+
+## Context
+
+Validator consensus key rotation feature has been discussed and requested for a long time, for the sake of safer validator 
+key management policy (e.g. https://github.com/tendermint/tendermint/issues/1136). So, we suggest one of the simplest form of
+validator consensus key rotation implementation mostly onto Cosmos-SDK.
+
+## Decision
+
+### Pseudo procedure for consensus key rotation
+
+- create new random consensus key.
+- create and broadcast a transaction with a `MsgRotateConsPubKey` that states the new consensus key is now coupled with the validator operator with signature from the validator's operator key.
+- old consensus key becomes unable to participate on consensus immediately after the update of key mapping state on-chain.
+- start validating with new consensus key.
+- validators using HSM and KMS should update the consensus key in HSM to use the new rotated key after the height `h` when `MsgRotateConsPubKey` committed to the blockchain.
+
+
+### Considerations
+
+- consensus key mapping information management strategy
+    - store history of each key mapping changes in the kvstore.
+    - the state machine can search corresponding consensus key paired with given validator operator for any arbitrary height in a recent unbonding period.
+    - the state machine does not need any historical mapping information which is past more than unbonding period.
+- limits
+    - a validator cannot rotate its consensus key more than N time for any unbonding period, to prevent spam.
+    - parameters can be decided by governance and stored in genesis file.
+- slash module
+    - slash module can search corresponding consensus key for any height so that it can decide which consensus key is supposed to be used for given height.
+- abci.ValidatorUpdate
+    - tendermint already has ability to change a consensus key by ABCI communication(`ValidatorUpdate`).
+    - validator consensus key update can be done via creating new + delete old by change the power to zero.
+    - therefore, we expect we even do not need to change tendermint codebase at all to implement this feature.
+
```

Comment:
what about charging validators for rotating a key as per @cwgoes comment?

```
However, frequent key rotations will make bisection less efficient, since fewer headers will be able to be skipped by the light client, so limiting the rate of key change is prudent. I think it might be sufficient to charge validators some nominal amount for changing their key, which could be directed to the community pool - that would "internalize the externality" of more expensive light client verification. Rate limiting to once per unbonding period or comparable duration is also probably fine.
```
https://github.com/tendermint/tendermint/issues/1136#issuecomment-542656685

Reply:
It was there in original suggestion but it was removed.
I am quite neutral to the charging.
But I think if we charge cost, I think the fee should be higher when number of rotation for given period is high.

- suggestion example : fee = 2 ^ (# of rotation for recent unbonding period) atoms
so, for 10th time in recent unbonding period, it costs 2^10 = 1024 atoms

Reply:
Agree re exponential cost.

Reply:
Yes, the idea is to have an on-chain parameter limiting the total (N) number of rotations per unbonding period. In addition, we can specifically charge higher gas costs for this specific Tx.

Reply:
we will update the spec as below
fee = initial_fee * 2^n 
(n = # of rotation for recent unbonding period, not including this one)
the initial_fee can be decided by governance and our suggestion is 1 atom.
Does it sound OK?

Reply:
Updated.